frontend/windows/handle_zip.go

```diff
@@ -0,0 +1,213 @@
+
+	def, err := st.Marshal(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error marshalling llb: %w", err)
+	}
+
+	res, err := client.Solve(ctx, gwclient.SolveRequest{
+		Definition: def.ToPB(),
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	ref, err := res.SingleRef()
+	// Do not return a nil image, it may cause a panic
+	return ref, &dalec.DockerImageSpec{}, err
+}
+
+func specToSourcesLLB(spec *dalec.Spec, sOpt dalec.SourceOpts, opts ...llb.ConstraintsOpt) (map[string]llb.State, error) {
+	out := make(map[string]llb.State, len(spec.Sources))
+	for k, src := range spec.Sources {
+		isDir := dalec.SourceIsDir(src)
+
+		s := ""
+		switch {
+		case src.DockerImage != nil:
+			s = src.DockerImage.Ref
+		case src.Git != nil:
+			s = src.Git.URL
+		case src.HTTP != nil:
+			s = src.HTTP.URL
+		case src.Context != nil:
+			s = src.Context.Name
+		case src.Build != nil:
+			s = fmt.Sprintf("%v", src.Build.Source)
+		case src.Inline != nil:
+			s = "inline"
+		default:
+			return nil, fmt.Errorf("no non-nil source provided")
+		}
+
+		pg := dalec.ProgressGroup("Add spec source: " + k + " " + s)
+		st, err := src.AsState(k, sOpt, append(opts, pg)...)
+		if err != nil {
+			return nil, err
+		}
+
+		if isDir {
+			st = llb.Scratch().File(llb.Copy(st, "/", filepath.Join("/", k)))
+		}
+
+		out[k] = st
+	}
+
+	return out, nil
+}
+
+func mapToArraySortedByKeys[T any](m map[string]T) []T {
+	keys := dalec.SortMapKeys(m)
+	out := make([]T, 0, len(keys))
+	for _, k := range keys {
+		out = append(out, m[k])
+	}
+	return out
+}
+
+func buildBinaries(spec *dalec.Spec, worker llb.State, sOpt dalec.SourceOpts) (llb.State, error) {
+	sources, err := specToSourcesLLB(spec, sOpt)
+	if err != nil {
+		return llb.Scratch(), err
+	}
+
+	patchedMap := dalec.PatchSources(worker, spec, sources)
+
+	patched := mapToArraySortedByKeys(patchedMap)
+
+	combined := dalec.MergeAtPath(llb.Scratch(), patched, "/")
+	buildScript := createBuildScript(spec)
+
+	binaries := maps.Keys(spec.Artifacts.Binaries)
+	script := generateInvocationScript(binaries)
+
+	buildDeps := spec.GetBuildDeps(targetKey)
+
+	work := worker
+	if len(buildDeps) > 0 {
+		slices.Sort(buildDeps)
+
+		work = work.Run(
+			shArgs("apt-get update && apt-get install -y "+strings.Join(buildDeps, " ")),
+			varCacheAptMount,
+			varLibAptMount,
+		).Root()
+	}
+
+	artifacts := work.Run(
+		shArgs(script.String()),
+		llb.Dir("/build"),
+		llb.AddMount("/build", combined),
+		llb.AddMount("/build/scripts", buildScript),
+		llb.Network(llb.NetModeNone),
+	).AddMount(outputDir, llb.Scratch())
+
+	return artifacts, nil
+}
+
+func getZipLLB(worker llb.State, name string, artifacts llb.State) llb.State {
+	outName := filepath.Join(outputDir, name+".zip")
+	zipped := worker.Run(
+		shArgs("zip "+outName+" *"),
+		llb.Dir("/tmp/artifacts"),
+		llb.AddMount("/tmp/artifacts", artifacts),
+	).AddMount(outputDir, llb.Scratch())
+	return zipped
+}
+
+func generateInvocationScript(binaries []string) *strings.Builder {
```

Comment:
any particular reason to do this as string script instead of building up LLB steps in this function?

Reply:
The script is just complicated enough (it loops over the spec's `artifacts.binaries`) to warrant a subroutine.